### ISSUE TITLE
Vendor by tag

### ISSUE THREAD

**Issue by onlyjob** [rsc/letsencrypt#15]
_xenolf/lego_ follows [semantic versioning](http://semver.org) practice therefore it should be vendored only by tag. Please do not vendor random unreleased snapshots and use [stable releases](https://github.com/xenolf/lego/releases) only.

Please update for current release of _xenolf/lego_ which is v0.3.1.

Thanks.

**Comment by onlyjob** [comment]
Updating _xenolf/lego_ to v0.3.1 would be very helpful in order to resolve FTBFS that is holding introduction of _rsc/letsencrypt_ to [Debian](http://debian.org).
[New release of Docker Distribution](https://github.com/docker/distribution/releases/tag/v2.5.0) just gained dependency on _rsc/letsencrypt_...

**Comment by aprice** [comment]
Libraries shouldn't vendor at all, it causes duplication of dependencies among other issues. Broad consensus in the Go community is that only binaries (package main) should vendor their dependencies, libraries should never use vendor. As soon as I see a "vendor" directory in a library project, I immediately write it off - I will never use it unless I want it badly enough to fork it and remove the vendoring.

**Comment by onlyjob** [comment]
By "vendor" I meant "depend" (just trying to use Golang terminology...) Any chance to update _letsencrypt_ for compatibility with xenolf/lego v0.3.1 ASAP?

**Comment by tonyhb** [comment]
Isn't this vendoring broke entirely? It states that it checks out `ca19a90028e242e878585941c2a27c8f3b3efc25`, though that commit doesn't actually contain the function `TLSSNI01ChallengeCertDomain` which this library uses.

See: https://github.com/xenolf/lego/blob/ca19a90028e242e878585941c2a27c8f3b3efc25/acme/tls_sni_challenge.go

vs the vendored:

https://github.com/rsc/letsencrypt/blob/master/vendor/github.com/xenolf/lego/acme/tls_sni_challenge.go#L44

**Comment by tonyhb** [comment]
I think this contains a commited dirty vendor state.

**Comment by vikstrous** [comment]
I can't find any version of lego that contains the string `TLSSNI01ChallengeCertDomain`. What version is vendored here?

Update: I found where the relevant function was added:

https://github.com/xenolf/lego/commit/617dd4d37cb9aae7d81bfef29f8c75067d27d70a#diff-a18bf20f679d5521e07174e6e72370dfR48

It's called TLSSNI01ChallengeCert. It was never called TLSSNI01ChallengeCertDomain.

**Comment by vikstrous** [comment]
Looks like this is already fixed in https://github.com/rsc/letsencrypt/pull/12

**Comment by rsc** [comment]
I modified the code after vendoring it. The vendor.json file records the revision I started with. The diffs are my local modifications. I prefer to record hashes because they don't change, while tags can.

**Comment by onlyjob** [comment]
Looks like you are doing a lot to make your software unmaintainable downstream, @rsc... :( :(
Normally I wouldn't bother but unfortunately _rsc/letsencrypt_  is a dependency of Docker Distribution which is broken due to this problem. I reckon I have to raise the issue with Docker... :(

**Comment by vikstrous** [comment]
There's already a fork and a PR on distribution. https://github.com/docker/distribution/pull/1792

**Comment by rsc** [comment]
A few points.
1. I've recorded sufficient history in vendor.json that vendoring tools should be able to deal with this situation. If they can't, then they should be fixed. I shouldn't have to make changes here because the vendoring tools are inadequate. I made the adjustments I needed in my copy of xenolf/lego. That's how open source works: we don't all move in lock step. I think it's very important that vendoring tools can handle this situation. Leaving the code this way is a small nudge to help that happen.
2. As a reminder, rsc.io/letsencrypt is, like all of rsc.io, a personal side project. In contrast to, say, the code in the main Go project from golang.org, rsc.io/letsencrypt is _unsupported_, _unreviewed_, _untested_ and for that matter now _unused_ (by me). I'm honestly surprised that Docker Distribution depends on this (security-critical) code for production uses. I added https://godoc.org/rsc.io/letsencrypt#hdr-Warning to the documentation a week ago to make the situation clearer.
3. Developers are responsible for the decisions they make about which code to pull in from the internet at large. If the code layout here somehow breaks Docker Distribution, then that's unfortunate, but the code has been like this from the first commit, before Docker chose to use it. Part of importing open source code is making whatever local changes are needed to fit into your code base. If Docker Distribution needs a change, Docker can easily make it in their copy. Again, that's how open source works: we don't all move in lock step. Like Dave Cheney [wisely said](https://twitter.com/davecheney/status/616931340466786304), "You don't download, or import, a software dependency, you adopt it. Like adopting pets, it's a responsibility for the life of your product." Docker Distribution took on the responsibility of using this unsupported, untested code. So yes, I reckon you do have to raise the issue with Docker.

I apologize for not making this admittedly trivial change. But the tooling and the processes here are both coming up short, and I believe it is important to the long-term health and scalability of the Go ecosystem to make sure that those get corrected.
